src/cdomains/apron/linearTwoVarEqualityDomain.apron.ml

```diff
@@ -0,0 +1,881 @@
+*)
+
+open Batteries
+open GoblintCil
+open Pretty
+module M = Messages
+open Apron
+open VectorMatrix
+
+
+
+module Mpqf = SharedFunctions.Mpqf
+
+module Equality = struct
+  (* (Some i, k) represents a sum of a variable with index i and the number k.
+     (None, k) represents the number k. *)
+  type t = (int option * Z.t) [@@deriving eq, ord, hash]
+  let zero = (None, Z.zero)
+  let var_zero i = (Some i, Z.zero)
+  let to_int x = Z.to_int @@ snd x
+  let print : t -> unit = fun (a, b) -> match a with
+    | None -> print_endline @@ "(None , " ^ Z.to_string b ^ ")"
+    | Some x -> print_endline @@ "(Some " ^ string_of_int x ^ ", " ^ Z.to_string b ^ ")"
+end
+
+module EqualitiesArray = struct
+  include Array
+  type t = Equality.t Array.t [@@deriving eq, ord]
+
+  let hash : t -> int = Array.fold_left (fun acc a -> 31 * acc + Equality.hash a) 0
+
+  let empty () = [||]
+
+  let make_empty_array len = Array.mapi (fun i (x, y) -> (Some i, Z.zero)) (make len Equality.zero)
+
+  let add_empty_column arr index = 
+    let num_vars = length arr in
+    if index > num_vars then failwith "n too large" else
+      let new_array = make (num_vars + 1) (Equality.var_zero index) in
+      if index = 0 
+      then blit arr 0 new_array 1 (num_vars - 1) 
+      else blit arr 0 new_array 0 index; 
+      if index <> num_vars 
+      then blit arr index new_array (index + 1) (num_vars - index);
+      new_array
+
+  let add_empty_columns m indexes = (** same as add_empty_columns for Matrix (see vectorMatrix.ml)*)
+    let nnc = length indexes in
+    if nnc = 0 then m else
+      let nc = length m in
+      let offset = ref 0 in
+      let offset_map = Array.init nc (fun j ->
+          while !offset < nnc && !offset + j = indexes.(!offset) do incr offset; done;
+          !offset)
+      in let add_offset_to_array_entry (var, offs) = 
+           Option.map (fun var_index -> var_index + offset_map.(var_index)) var, offs in
+      let m' = make_empty_array (nc + nnc) 
+      in Array.iteri (fun j eq -> m'.(j + offset_map.(j)) <- add_offset_to_array_entry eq) m;
+      m'
+
+  let del_cols m indexes =
+    let nrc = length indexes in
+    if nrc = 0 || length m = 0 then m
+    else
+      let nc = length m in
+      if nc = nrc then [||] else
+        let offset = ref 0 in
+        let offset_map = Array.init nc (fun j ->
+            if indexes.(!offset) = j then (incr offset; 0) else !offset)
+        in let remove_offset_from_array_entry (var, offs) = 
+             Option.map (fun var_index -> var_index - offset_map.(var_index)) var, offs in
+        Array.init (nc - nrc) (fun j -> remove_offset_from_array_entry m.(j + offset_map.(j)))
+
+  let del_cols m cols = timing_wrap "del_cols" (del_cols m) cols
+
+  let is_empty m = length m = 0
+
+  let is_top_array m = Array.fold_lefti (fun b i (a, e) -> if Z.(e = zero) && Option.is_some a && Option.get a = i then b else false) true m
```

Comment:
If you find yourself doing `for_all_i` or `exists_i` several times it might make sense to define a helper for it.

Reply:
Y

Reply:
If you need to check `for_alli_i` or `exists_i` in several places, you could write functions taking an array and a predicate `int -> (int option, Z.t)` that abstract away the internal things such as the accumulator etc.

If you have these contained, you can then even think about doing a faster but more ugly implementation (e.g. doing shortcuts), as this ugliness would be local and wouldn't spread to everywhere.

Of course this makes only sense if you have many places where you could use these helpers.

Reply:
Ok sure, I will watch out for opportunities.